test/pipelinerun_test.go

```diff
@@ -198,6 +199,101 @@ func TestPipelineRun(t *testing.T) {
 	}
 }
 
+func TestPipelineRunStaysFinished(t *testing.T) {
+	logger := getContextLogger(t.Name())
+	client, namespace := setup(t, logger)
+	index := 0
+
+	knativetest.CleanupOnInterrupt(func() { tearDown(t, logger, client, namespace) }, logger)
+	defer tearDown(t, logger, client, namespace)
+
+	logger.Infof("Setting up test resources for TestPipelineRunStaysFinished test in namespace %s", namespace)
+	if _, err := client.KubeClient.Kube.CoreV1().Secrets(namespace).Create(getPipelineRunSecret(index, namespace)); err != nil {
+		t.Fatalf("Failed to create secret `%s`: %s", getName(hwSecret, index), err)
+	}
+
+	if _, err := client.KubeClient.Kube.CoreV1().ServiceAccounts(namespace).Create(getPipelineRunServiceAccount(index, namespace)); err != nil {
+		t.Fatalf("Failed to create SA `%s`: %s", getName(hwSA, index), err)
+	}
+
+	if _, err := client.TaskClient.Create(&v1alpha1.Task{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      getName(hwTaskName, index),
+		},
+		Spec: v1alpha1.TaskSpec{
+			// Reference build: https://github.com/knative/build/tree/master/test/docker-basic
+			Steps: []corev1.Container{{
+				Name:  "say-hello",
+				Image: "ubuntu",
+				// Private docker image for Build CRD testing
+				Command: []string{"/bin/bash"},
+				Args:    []string{"-c", "echo hello"},
+			}},
+		},
+	}); err != nil {
+		t.Fatalf("Failed to create Task `%s`: %s", getName(hwTaskName, index), err)
+	}
+
+	if _, err := client.PipelineClient.Create(getHelloWorldPipelineWithSingularTask(index, namespace)); err != nil {
+		t.Fatalf("Failed to create Pipeline `%s`: %s", getName(hwPipelineName, index), err)
+	}
+
+	prName := fmt.Sprintf("%s%d", hwPipelineRunName, index)
+	if _, err := client.PipelineRunClient.Create(getHelloWorldPipelineRun(index, namespace)); err != nil {
+		t.Fatalf("Failed to create PipelineRun `%s`: %s", prName, err)
+	}
+
+	logger.Infof("Waiting for PipelineRun %s in namespace %s to complete", prName, namespace)
+	if err := WaitForPipelineRunState(client, prName, pipelineRunTimeout, func(tr *v1alpha1.PipelineRun) (bool, error) {
+		condition := tr.Status.GetCondition(duckv1alpha1.ConditionSucceeded)
+		if condition != nil {
+			if condition.IsTrue() {
+				return true, nil
+			} else if condition.IsFalse() {
+				return true, fmt.Errorf("Pipeline run %s has failed with status %v", prName, condition.Status)
+			}
+		}
+		return false, nil
+	}, "PipelineRunSuccess"); err != nil {
+		t.Fatalf("Error waiting for PipelineRun %s to finish: %s", prName, err)
+	}
+
+	taskRunName := strings.Join([]string{prName, hwPipelineTaskName1}, "-")
+
+	logger.Infof("Deleting TaskRun and waiting for reconciliation")
+	if err := client.TaskRunClient.Delete(taskRunName, &metav1.DeleteOptions{}); err != nil {
+		t.Fatalf("Could not delete TaskRun %s: %s", taskRunName, err)
+	}
```

Comment:
one idea that would cut down on the number of e2e tests we have is to add the logic to delete the `TaskRun` to the existing test - I think that'd be fine to do - we want unit tests that only test one thing, but for end to end tests id really rather have a just a couple more complex scenarios

Reply:
I thought about that, but couldn't figure out a non-awkward way to wedge that into the existing loop of multiple test cases using the same execution cleanly. I'll loop back on it.

Reply:
> For example triggering reconcile on already successful pipelinerun should not generate kubernetes action objects would be a test case for this.

Oh yeah that makes a lot of sense thanks @shashwathi !

I think that's what you were looking for earlier @abayer

Reply:
It most definitely is. Thanks!